sys/linux/socket_netlink_route.txt

```diff
@@ -56,6 +57,9 @@ netlink_msg_route [
 	newlinkprop		netlink_msg[RTM_NEWLINKPROP, ifinfomsg[AF_UNSPEC], ifla_policy]
 #	RTM_GETLINKPROP is currently not used by the kernel.
 	dellinkprop		netlink_msg[RTM_DELLINKPROP, ifinfomsg[AF_UNSPEC], ifla_policy]
+	new_nexthop		netlink_msg[RTM_NEWNEXTHOP, nhmsg[AF_UNSPEC], rtm_nh_policy]
+	del_nexthop		netlink_msg[RTM_DELNEXTHOP, nhmsg[AF_UNSPEC], rtm_nh_policy]
+	get_nexthop		netlink_msg[RTM_GETNEXTHOP, nhmsg[AF_UNSPEC], rtm_nh_policy]
```

Comment:
For `del_nexthop`, it seems that only `NHA_ID` is valid: https://elixir.bootlin.com/linux/v5.8-rc4/source/net/ipv4/nexthop.c#L1640

____

For `get_nexthop`, depending on the netlink_msg flags (do vs dump), `NHA_ID`, `NHA_OIF`, `NHA_MASTER`, `NHA_GROUPS`, `NHA_FDB` are the valid options:
- https://elixir.bootlin.com/linux/v5.8-rc4/source/net/ipv4/nexthop.c#L1640
- https://elixir.bootlin.com/linux/v5.8-rc4/source/net/ipv4/nexthop.c#L1786

____

For `new_nexthop`, `NHA_GROUPS` and `NHA_MASTER` are invalid: https://elixir.bootlin.com/linux/v5.8-rc4/source/net/ipv4/nexthop.c#L1455
____

We might want to have seperate policies for each to conform with the valid attributes.

Reply:
Yes, but isn't this how it is done overall?
Eg, ifla_policy:
https://github.com/google/syzkaller/blob/b712935571355df44e126c5b27c98ccd53d1d18b/sys/linux/socket_netlink_route.txt#L199
Which here (and in most other places) has an exit for when there are to many:
https://elixir.bootlin.com/linux/v5.8-rc4/source/net/core/rtnetlink.c#L3435

Should then, taking ifla_policy as an example, be actually separated into ifla_(dump|set|get)_policy instead?
Or should I keep this one as is?

Reply:
One reason might be that they haven't seen much value in introducing this complexity in the descriptions.

Another can be that there is considerable amount of code being run until the error point, thus, it might be useful to include them to see if they trigger any bugs such as something not being cleaned up if the error occurs.

There are many such examples as you mentioned. Let's leave it as is and hear about this from other reviewers.

Reply:
There is no one true answer for keeping it closer to the kernel definitions and simpler vs more complex and efficient.
We did it the other way around too:
https://github.com/google/syzkaller/blob/b712935571355df44e126c5b27c98ccd53d1d18b/sys/linux/socket_netlink_netfilter_ipset.txt#L63-L156
I believe here we split a single kernel policy into per-command policies.

I am mildly leaning towards a more specialized policies.
Please keep the naming convention as rtm_nh_policy_some_suffix.

Reply:
I split them to specialized policies, let me know if I should keep it like this (more specialized) or merge the ones for a simple entry point (i.e., `rtm_nh_policy_new`, `rtm_nh_policy_new_fdb`, and `rtm_nh_policy_new_blackhole` into only `rtm_nh_policy_new`).